File: elixir/medium/0948/0948.exs
defmodule Solution do
  @spec bag_of_tokens_score(tokens :: [integer], power :: integer) :: integer
  def bag_of_tokens_score(tokens, power) do
    tokens_tuple =
      tokens
      |> Enum.sort()
      |> List.to_tuple()

    do_score(tokens_tuple, power, 0, 0, tuple_size(tokens_tuple) - 1)
  end

  defp do_score(tokens, power, score, l, r)
       when l <= r and elem(tokens, l) <= power do
    do_score(tokens, power - elem(tokens, l), score + 1, l + 1, r)
  end

  defp do_score(tokens, power, score, l, r)
       when l < r and score >= 1 do
    do_score(tokens, power + elem(tokens, r), score - 1, l, r - 1)
  end

  defp do_score(_tokens, _power, score, _l, _r), do: score
end

ExUnit.start()

defmodule AssertionTest do
  use ExUnit.Case, async: true

  test "test bag_of_tokens_score" do
    assert Solution.bag_of_tokens_score([100], 50) == 0
    assert Solution.bag_of_tokens_score([200, 100], 150) == 1
    assert Solution.bag_of_tokens_score([100, 200, 300, 400], 200) == 2
  end
end

File: py/medium/0948/0948.py
from collections import deque
from typing import List


class Solution:
    def bagOfTokensScore(self, tokens: List[int], power: int) -> int:
        score = 0
        sorted_tokens = deque(sorted(tokens))
        l, r = 0, len(sorted_tokens) - 1

        while True:
            while l <= r and sorted_tokens[l] <= power:
                power -= sorted_tokens[l]
                l += 1
                score += 1
            if score >= 1 and l < r:
                power += sorted_tokens[r]
                r -= 1
                score -= 1
            else:
                break

        return score


solution = Solution()
assert solution.bagOfTokensScore([100], 50) == 0
assert solution.bagOfTokensScore([200, 100], 150) == 1
assert solution.bagOfTokensScore([100, 200, 300, 400], 200) == 2
